compile/api/src/main/scala/sbt/ClassToAPI.scala

```diff
@@ -127,14 +135,48 @@ object ClassToAPI {
   def upperBounds(ts: Array[Type]): api.Type =
     new api.Structure(lzy(types(ts)), lzyEmptyDefArray, lzyEmptyDefArray)
 
-  def fieldToDef(enclPkg: Option[String])(f: Field): api.FieldLike =
+  @deprecated("Use fieldToDef[4] instead", "0.13.9")
+  def fieldToDef(enclPkg: Option[String])(f: Field): api.FieldLike = {
+    val c = f.getDeclaringClass()
+    fieldToDef(c, classFileForClass(c), enclPkg)(f)
+  }
+
+  def fieldToDef(c: Class[_], _cf: => ClassFile, enclPkg: Option[String])(f: Field): api.FieldLike =
     {
       val name = f.getName
       val accs = access(f.getModifiers, enclPkg)
       val mods = modifiers(f.getModifiers)
       val annots = annotations(f.getDeclaredAnnotations)
-      val tpe = reference(returnType(f))
-      if (mods.isFinal) new api.Val(tpe, name, accs, mods, annots) else new api.Var(tpe, name, accs, mods, annots)
+      // generate a more specific type for constant fields
+      val specificTpe: Option[api.Type] =
+        if (mods.isFinal) {
+          val cf = _cf
+          val attributeInfos = cf.fields.find(_.name.exists(_ == name)).toSeq.flatMap(_.attributes)
+          // create a singleton type ending with the ConstantValue of this field. because this type
+          // is purely synthetic, it's fine that the name might contain filename-banned characters.
+          attributeInfos.collectFirst {
+            case ai @ classfile.AttributeInfo(Some("ConstantValue"), _) =>
+              try {
+                c.getName.split("\\.").toSeq :+ (name + "$" + cf.constantValue(ai))
```

Comment:
> I think your encoding makes sense, but it does not capture the type of the field, which is unlikely not to also result in a change of value when it changes, but it could (and thus changes to the type of the field would go undetected).
Aha! Absolutely right. Will fix.